### DISCUSSION TITLE
Do we need to build JARs for UI Customisation on Keycloak 21.1.1

### DISCUSSION THREAD

**Discussion by gautamjgithub** [keycloak/keycloak#21314] [Q&A]
For Following theme changes:

1) changing background image, adding new logo on login. register and other screens
2) Adding custom fields with javascript validations on Register flow
3) Adding OTP authentication flow

For any of the above change, Do we need to modify jar file as it was done earlier in Keycloak 16 or from now onwards it is all managed through configuration.

**Comment by dteleguin** [comment]
Generally, you won't need to modify any Keycloak JARs in order to deploy custom themes.

With newer Keycloak versions, you can either copy your theme to the `themes` directory, or deploy it as a provider JAR. See [documentation](https://www.keycloak.org/docs/latest/server_development/index.html#deploying-themes) for more info.

**Reply by gautamjgithub** [reply]
Thanks @dteleguin . Themes is fine. Regarding sending the register user information to an external API, which SPI has to be built?

Is there any existing JAR which can be extracted, modified and once again jar can be created?

**Reply by dteleguin** [reply]
If you want to send the newly registered user's info to an external API, I'd recommend implementing an [event listener](https://www.keycloak.org/docs/latest/server_development/#_events) and listen for events of type `REGISTER`. This will need to be packaged as a separate JAR and deployed into the  `providers` directory (no Keycloak JAR modification needed).
